### PR TITLE
House: Chaining the models City Country

### PR DESCRIPTION
**Pull request by shlior7**
Relevant Issues: #114
I removed the countries selection logic from the HTML file, and replace it by the lib "[django-smart-select](https://django-smart-selects.readthedocs.io/en/latest/)" 

this lib chains models with jquery fetching automatically in the admin and in any form as seen in the docs.

the latest version of this lib[(1.6.0)](https://github.com/jazzband/django-smart-selects/blob/master/CHANGES.rst) is the first supporting python 3.9 which we are using but this version isn't released yet. so the solution i saw people use is to explicitly state that we are using the lib from the version of the repo in github see Pipfile.

and thus no more worrying about logic in the HTML.

Because all this magic is happening from the 3rd party library there is no real code to test that isn't tested thoroughly in the library itself, so I don't think any new tests need to be created If there is I would like to hear the idea of the test because the output of this PR is still "frontendy" meaning changes are happening to the HTML element "select" and its options so I cant think of a test that tests the code behind that. But I checked this on the House filter form and in the admin and this works perfectly 

The tests changed only because we're creating the initial countries and cities data using the django command loaddata, so the tests need to run that first as a fixture. and that created some problems with the test classes so I moved the problematic ones to be functional and that fixed the problem.

https://user-images.githubusercontent.com/34008861/168373694-0c1ed856-e53a-4873-9f27-9e774aed67c8.mov

https://user-images.githubusercontent.com/34008861/168373920-ef21610d-0757-494f-ad58-f90a673014f5.mov

### CODE REVIEW

File: expenses/migrations/0001_initial.py [hunk 1]
```diff
@@ -20,9 +20,34 @@ class Migration(migrations.Migration):
             fields=[
                 ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                 ('description', models.TextField(default='', max_length=250)),
-                ('category', models.CharField(choices=[('Rent', 'Rent'), ('Mortgage', 'Mortgage'), ('Bills', 'Bills'), ('Transportation', 'Transportation'), ('Clothing', 'Clothing'), ('Healthcare', 'Healthcare'), ('Food', 'Food'), ('Insurance', 'Insurance'), ('Kids', 'Kids'), ('Culture', 'Culture'), ('Vacations', 'Vacations'), ('Other', 'Other')], default='Other', max_length=32)),
-                ('amount', models.IntegerField(blank=True, null=True, validators=[django.core.validators.MinValueValidator(0)])),
-                ('date', models.DateTimeField(default=django.utils.timezone.now)),
+                (
+                    'category',
+                    models.CharField(
```
Comment (ShellyGolden) [inline expenses/migrations/0001_initial.py:25]: I think that line 25 can be on line 24.
The context is more understandable, at least for me

File: expenses/migrations/0001_initial.py [hunk 2]
```diff
@@ -20,9 +20,34 @@ class Migration(migrations.Migration):
...
+                            ('Vacations', 'Vacations'),
+                            ('Other', 'Other'),
+                        ],
+                        default='Other',
+                        max_length=32,
+                    ),
+                ),
+                (
+                    'amount',
+                    models.IntegerField(
```
Comment (ShellyGolden) [inline expenses/migrations/0001_initial.py:46]: also here - like 24

File: house/fixtures/countries_cities.json [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    },
+    {
+        "fields": {
+            "country": 5,
+            "name": "Cambridge"
+        },
+        "model": "house.city",
+        "pk": 26
+    }
+]
```
Comment (ShellyGolden) [inline house/fixtures/countries_cities.json:245]: Add a new line

File: house/migrations/0001_initial.py [hunk 4]
```diff
@@ -34,13 +34,80 @@ class Migration(migrations.Migration):
...
+                            ('Other', 'Other'),
+                            ('Unemployed', 'Unemployed'),
+                        ],
+                        default='Other',
+                        max_length=50,
+                    ),
+                ),
+                (
+                    'income',
+                    models.IntegerField(
```
Comment (ShellyGolden) [inline house/migrations/0001_initial.py:85]: the same comment like the other migrations file :)

File: house/migrations/0001_initial.py [hunk 5]
```diff
@@ -34,13 +34,80 @@ class Migration(migrations.Migration):
...
+                        blank=True, null=True, validators=[django.core.validators.MinValueValidator(0)]
+                    ),
+                ),
+                (
+                    'children',
+                    models.IntegerField(
```
Comment (ShellyGolden) [inline house/migrations/0001_initial.py:91]: also here

File: scripts/countries_cities_data.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        "name": "United Kingdom",
+        "cities": [
+            "London",
+            "Huddersfield",
+            "Glasgow",
+            "Edinburgh",
+            "Cambridge"
+        ]
+    }
+]
```
Comment (ShellyGolden) [inline scripts/countries_cities_data.json:53]: Add a new line

File: tests/house_tests.py [hunk 7]
```diff
@@ -126,43 +134,52 @@ def form_data_filter_tests():
...
-            get_template(MINE_MINE_PAGE_ROUTE)
-        except TemplateDoesNotExist:
-            assert False, f"Template {MINE_MINE_PAGE_ROUTE} does not exist"
-
-    def test_mine_sidebar_views_templates(self):
-        try:
-            get_template(MINE_SIDEBAR_ROUTE)
-        except TemplateDoesNotExist:
-            assert False, f"Template {MINE_SIDEBAR_ROUTE} does not exist"
+@pytest.fixture
```
Comment (ShellyGolden) [inline tests/house_tests.py:137]: Why are these tests within this PR?
I don't see that you mentioned that the PR is blocked by another PR

File: tips/models.py [hunk 8]
```diff
@@ -8,20 +8,11 @@ class Tip(models.Model):
     text = models.TextField(max_length=200)
     date = models.DateTimeField(default=timezone.now)
 
-    category = models.CharField(
-         max_length=32,
-         choices=Expenses.Category.choices,
-         default=Expenses.Category.OTHER
-    )
+    category = models.CharField(max_length=32, choices=Expenses.Category.choices, default=Expenses.Category.OTHER)
```
Comment (ShellyGolden) [inline tips/models.py:11]: Why did you change this model? 
He is not tied to the house

**Review by ShellyGolden — COMMENTED** [review]
I left a few comments 
good job :)

File: scripts/load_countries_cities.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import json
+
+
+def run():
+    pwd = input("Write the countries cities data json file path from root repo: ")
+    pwd = "./scripts/countries_cities_data.json" if pwd == "" else pwd
```
Comment (LiorKGOW) [inline scripts/load_countries_cities.py:6]: What is the meaning of the name pwd? maybe use a name like: file_path? :)

File: scripts/load_countries_cities.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    with open(pwd, "r") as f:
```
Comment (LiorKGOW) [inline scripts/load_countries_cities.py:8]: Could you give a more meaningful name for f?
like file? json_file? data_file? or maybe coutry_city_data?

File: tests/house_tests.py [hunk 11]
```diff
@@ -9,99 +9,107 @@
...
-            city=City.objects.get(name='Tel Aviv'),
-            income=5000,
-            children=1,
-            parent_profession_1=Job.TEACHER,
-        ).save()
-        houses = _filter_houses_by_form(form_data_filter_tests(), House.objects.all())
-        assert len(houses) == 2
-        assert House.objects.get(parent_profession_1=Job.TEACHER) in houses
-        assert House.objects.get(parent_profession_2=Job.TEACHER) in houses
+def test_save_house(db):
```
Comment (LiorKGOW) [inline tests/house_tests.py:12]: Shouldn't these test be included in a proper Test Class?
Like we did :
@pytest.mark.django_db
class TestFilterHouseForm:
   def test1
   def test 2
   ...

File: tips/models.py (hunk 8 above)
Comment (LiorKGOW) [reply tips/models.py:11]: Lior just rearanged this line, The data stayed the same :)

File: house/models.py [hunk 12]
```diff
@@ -49,8 +50,15 @@ class House(models.Model):
     house_id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
     name = models.CharField(max_length=50)
     public = models.BooleanField(default=False)
-    country = models.ForeignKey(Country, on_delete=models.SET_NULL, null=True)
-    city = models.ForeignKey(City, on_delete=models.SET_NULL, null=True)
+    country = models.ForeignKey(Country, on_delete=models.CASCADE, null=True)
+    city = ChainedForeignKey(
```
Comment (LiorKGOW) [inline house/models.py:54]: Nice Solution to the problem
i think that it is better then what I tried to do
I tried to fix the country-city problem by Implementing a Dependent/Chained Dropdown List with Django
and adding it to the django form creating a house, but i haven't managed to make it to work...
Good Job Lior 👍

**Review by LiorKGOW — COMMENTED** [review]
Nice Solution to the problem @shlior7 
i think that it is better then what I tried to do
I tried to fix the country-city dependency problem by Implementing a Dependent/Chained Dropdown List with Django
and adding it to the django form of creating a house, but i haven't managed to make it to work...
Good Job Lior 👍 
I had a few comments please take a look :)

File: expenses/migrations/0001_initial.py (hunk 1 above)
Comment (shlior7) [reply expenses/migrations/0001_initial.py:25]: This is what my formatter formatted. I think its understandable but anyway its a migration and generated automatically so this is not that critical.

File: expenses/migrations/0001_initial.py (hunk 2 above)
Comment (shlior7) [reply expenses/migrations/0001_initial.py:46]: same here

File: house/models.py (hunk 12 above)
Comment (shlior7) [reply house/models.py:54]: Thanks Lior, this was a bit of work mainly because of no support for our version of Django and python and the way we initiated the countries and cities data. so this comment is much appreciated 🥇

File: tests/house_tests.py (hunk 11 above)
Comment (shlior7) [reply tests/house_tests.py:12]: So no, there is no need and actually its better to use Pytest functionally. some classes that use the country and cities didnt work after I changed the way we load the initial data, so they had to migrate to be functional tests to use the loaddata fixture that makes sure it loads the initial data before the tests

File: tests/house_tests.py (hunk 7 above)
Comment (shlior7) [reply tests/house_tests.py:137]: I branched from the main branch so I don't think this PR is blocked by any other PR. maybe I don't understand the question

File: house/admin.py [hunk 13]
```diff
@@ -1,5 +1,20 @@
 from django.contrib import admin
 
-from house.models import House
+from house.models import House, City, Country
 
-admin.site.register(House)
+
+class CountryAdmin(admin.ModelAdmin):
+    list_display = ["name"]
```
[the remark this answers is not in the document]
Comment (shlior7) [reply house/admin.py:7]: it needs to be in a list because the only field is "name" the tuple doesnt apply, it could be a list at HouseAdmin and CityAdmin (a tuple is a list behind the curtains shh.... )

File: house/migrations/0001_initial.py (hunk 4 above)
Comment (shlior7) [reply house/migrations/0001_initial.py:71]: I agree, lint comments on migration files are irrelevant.

**Review by LiorKGOW — APPROVED** [review]

**Review by ronyyosef — CHANGES_REQUESTED** [review]
Please do not use formater on the migrations files
It makes it hard for us to understand what is the difference

**Comment by shlior7** [comment]
> there are two commits, please rebase and fix typo in title and commit-msg

squash them?

**Comment by adiabramovitch** [comment]
> > there are two commits, please rebase and fix typo in title and commit-msg
> 
> 
> 
> what typo?

Did you me to chain (chaining)? Or change (changing)?

**Comment by ShellyGolden** [comment]
Is this PR based on other PRs?
If so - on which? I would love for you to point out above
If not - why does it show that there are updates on all sorts of things that are not relevant to PR? For example - tests / house_tests.py
@shlior7

**Comment by shlior7** [comment]
> Is this PR based on other PRs? If so - on which? I would love for you to point out above If not - why does it show that there are updates on all sorts of things that are not relevant to PR? For example - tests / house_tests.py @shlior7

_The tests changed only because we're creating the initial countries and cities data using the django command loaddata, so the tests need to run that first as a fixture. and that created some problems with the test classes so I moved the problematic ones to be functional and that fixed the problem._

this PR is not blocked by any other PR, about the tests, please read the description.
but in short the way we load the cities and countries has changed for this PR to work and that triggered some changes in the tests as I explained

**Review by ronyyosef — APPROVED** [review]

**Review by Yoavsc129 — APPROVED** [review]

File: expenses/migrations/0001_initial.py [hunk 14]
```diff
@@ -11,7 +11,7 @@ class Migration(migrations.Migration):
     initial = True
 
     dependencies = [
-        ('house', '0002_add_initial_data'),
+        ('house', '__first__'),
```
Comment (LiorKGOW) [inline expenses/migrations/0001_initial.py:14]: What does this change do?
what does __first__ mean?
I have read here that it refers to the *first* migration. (and the change says it was 0002)
https://code.djangoproject.com/ticket/22861

**Review by LiorKGOW — APPROVED** [review]
Hey Lior
I had a question about one of the migration files, other then that I approve 👍

File: expenses/migrations/0001_initial.py (hunk 14 above)
Comment (shlior7) [reply expenses/migrations/0001_initial.py]: what the `__first__` means is that this migration needs to happen after the first migration of house.
because I removed the previous 0002 migration I noticed this expenses migration doesn't needs(depends) on 0002_add_initial_data(adds the cities and countries) but only needs the first house migration (0001_initial).
for readability and specifics I changed it to depend on `0001_initial`, but it works the same. hope this is more clear now :)

**Comment by svatares** [comment]
@shlior7 please resolve conflicts

File: expenses/migrations/0001_initial.py (hunk 14 above)
Comment (LiorKGOW) [reply expenses/migrations/0001_initial.py]: It is, at least for me
Thanks!

**Review by LiorKGOW — APPROVED** [review]

**Comment by ronyyosef** [comment]
Can you add the chaining also to the create house functionality?

**Comment by shlior7** [comment]
> Can you add the chaining also to the create house functionality?

Done, please take a look at the `create_house.html` for the changes

File: static/css/app_layout.css [hunk 15]
```diff
@@ -63,4 +63,4 @@ body {
 
 .fw-bolder {
   font-weight: 500 !important;
-}
+}
```
Comment (adiabramovitch) [inline static/css/app_layout.css:66]: new line missing

File: static/css/house_create.css [hunk 16]
```diff
@@ -1,9 +1,13 @@
...
 }
 
-#submit-button{
+#submit-button {
   width: 100%;
 }
+
+select {
+  display: block !important;
+}
```
Comment (adiabramovitch) [inline static/css/house_create.css:13]: ditto

**Comment by adiabramovitch** [comment]
I'll merge that PR as soon as you fix the new line issue. 
**But** a PR changing 28 files is **NOT** fine. 
it was supposed to be split into at least 5 different PRs - one changing the expense model, one changing the house model, and even making the backend and the froNT separated. 
It is not possible to track all the changes that occurred and pay attention to different problems that may occur.
